includes/classes/prod_cat_pulldown.php

```diff
@@ -0,0 +1,186 @@
+            LEFT JOIN " . TABLE_PRODUCTS_TO_CATEGORIES . " ptoc on (c.categories_id = ptoc.categories_id) "
+				. $this->attributes_join;
+
+			$this->sort = " ORDER BY categories_name";
+			$this->keyword_search_fields = [
+				'cd.categories_name',
+				'c.parent_id',
+				'cd.categories_description',
+				'c.categories_id',
+			];
+		}
+
+		private function product_vars()
+		{
+			$this->keyword_search_fields = [
+				'pd.products_name',
+				'p.products_model',
+				'pd.products_description',
+				'p.products_id',
+			];
+
+			$parm_2 = '%2$s ';
+			if (!$this->show_model) {
+				$parm_2 = '';
+			}
+
+			$this->sort = ' ORDER BY pd.products_name';
+			$this->output_string = '%1$s ' . $parm_2 . '(%3$s)';// format string with name first
+
+			if ($this->order_by == 'model') {
+				$this->sort = ' ORDER by p.products_model';
+				$this->output_string = (!empty($parm_2) ? $parm_2 . '-' : '') . ' %1$s (%3$s)'; // format string with model first
+			}
+
+			$this->sql = "SELECT p.products_id as common_id, pd.products_id, p.products_sort_order, p.products_price, p.products_model, ptc.categories_id
+                FROM " . TABLE_PRODUCTS . " p
+                LEFT JOIN " . TABLE_PRODUCTS_DESCRIPTION . " pd USING (products_id)
+                LEFT JOIN " . TABLE_PRODUCTS_TO_CATEGORIES . " ptc USING (products_id) "
+				. $this->attributes_join .
+				" WHERE pd.language_id = " . (int)$_SESSION['languages_id'];
+		}
+
+		private function run_sql()
+		{
+			global $db, $currencies;
+
+			$search = '';
+			if (!empty($this->keywords)) {
+				$search = zen_build_keyword_where_clause($this->keyword_search_fields,
+					trim(zen_db_input(zen_db_prepare_input($this->keywords))));
+			}
+			$this->sql .= $search;
+
+			switch (true) {
+				case ($this->option_filter === -1): // no selection made: do not list any products
+				case ($this->option_filter === ''): // no selection made: do not list any categories
+					// no selection made yet
+					break;
+				case ((int)$this->option_filter > 0) : // an Option Name was selected: show only products using attributes with this Option Name
+					$this->sql .= $this->condition;
+					$this->sql .= $this->sort;
+					break;
+				default: //legacy: show all products with attributes
+					$this->sql .= $this->sort;
+					break;
+			}
+
+			$results = $db->Execute($this->sql);
+
+			foreach ($results as $result) {
+				if (!in_array($result['common_id'], $this->exclude)) {
+					switch (true) {
+						case ($this->pulldown == 'category'):
+							$this->values[] = [
+								'id' => $result['categories_id'],
+								'text' => $this->category_text($result),
+							];
+							break;
+						default:
+							$display_price = zen_get_products_base_price($result['products_id']);
+							$name = zen_get_products_name($result['products_id']);
+							$this->values[] = [
+								'id' => $result['products_id'],
+								'text' => sprintf($this->output_string, trim(zen_clean_html($name)),
```

Comment:
If a space existed before or after the name, wouldn't that have been on purpose and therefore desired? As a result of this operation doing a trim, two names may appear in the same format though as unique values, further there may be code down stream that uses/processes any additional content that is in the `$name`; however, it is removed from this data here.  The code in `zen_draw_pull_down_menu` has two things, one a notifier allowing an alternate output result and also provides a level of the `zen_clean_html` function by replacement of common characters that would cause a problem with the html equivalent.

Basically wondering why the `$name` is being treated so early in processing? Note that similar management is not done on `products_model` nor when dealing with categories, the `categories_name`.

Reply:
with regards to spaces before and/or after a product name, YAGNI.  such an edge case; i can not think of any use case scenario where a store owner has product A and product B, and the name of said products is only differentiated by a space before or a space after the name.

i decided to do a trim on the name after the `zen_clean_html` as in testing on a different related project, the `zen_clean_html` function left some white space at the beginning (or end... i do not remember).

in looking at this, it is possible that i would need to clean the category name as well.  i may have missed that.  i have clients that embed html in their category name for display purposes, and i would have to look at results from their dataset.  as far as cleaning the products model.... perhaps, again an edge case, are there store owners that embed html in model number?

Reply:
So, the flip side to this though is that the `pull_down` method calls `zen_draw_pull_down_menu`. That function has its own notifier controls to support editing the content; however, the "restricted" display provided here is not included elsewhere in the Zen Cart flow path, thus allowing "unique" data to be present/incorporated/handled.  I mean sure if dead set on removing things like that by using `zen_clean_html`, then would highly recommend incorporating ways that those who actually want the additional content can work "with" this. (Ie. appropriately placed notifiers so that the data may be captured/updated.  Considering so much of this data is being stored in the class, there is really little that has to be passed by the notifier because most of it will already be captured in the class information. May still need to pass something, but...)

Yeah, depending on what else was in that content, the cleaning of the html could leave behind a space at either beginning or end.  My point though is that in general store owners/operators take action to make things look a specific way or contain specific information (those beginning will have a learning curve of some sort, regardless the actual output) and seems that when/if adding/changing things that it still be possible to get the same results even if in a better way and perhaps through more hoops.

As to data control of including/excluding an additional "outer" space, I see that its possible and could be done on purpose and while I don't use it, I don't see why it has to be forced to be removed.